### PR TITLE
Try to build with OCaml 4.08 version

### PR DESCRIPTION
**Pull request by XVilka**
With the new OCaml 4.08 release there are lot of nice optimizations and features out of the box, even without changing the current code. Worth to make sure it builds. For now lets see if everything passes OK.

### CODE REVIEW

**Comment by XVilka** [comment]
Strange error on 4.07 compiler:
```
- ocamldoc -html -d ../autodoc  -t "FrontC" cabs.ml cxml.ml cprint.ml ctoxml.ml frontc.ml
- ../autodoc/style.css: No such file or directory
- Fatal error: exception Failure("../autodoc/Cabs.html: No such file or directory")
- make[1]: *** [doc] Error 2
```
Also I don't get what is happening there - error message looks empty
```
<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+ /home/travis/.opam/opam-init/hooks/sandbox.sh "build" "./configure" "--prefix=/home/travis/.opam/4.07.0" "--with-cxx=`which clang++`" "--with-llvm-version=5.0.0git-929163d" "--with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config" "--mandir=/home/travis/.opam/4.07.0/man" "--enable-everything" "--enable-tests" "--disable-ida" "--disable-fsi-benchmark" "--disable-objdump" (CWD=/home/travis/.opam/4.07.0/.opam-switch/build/bap.master)
[ERROR] The compilation of bap failed at
        "/home/travis/.opam/opam-init/hooks/sandbox.sh build ./configure
        --prefix=/home/travis/.opam/4.07.0 --with-cxx=`which clang++`
        --with-llvm-version=5.0.0git-929163d
        --with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config
        --mandir=/home/travis/.opam/4.07.0/man --enable-everything
        --enable-tests --disable-ida --disable-fsi-benchmark
        --disable-objdump".
#=== ERROR while compiling bap.master =========================================#
# context     2.0.4 | linux/x86_64 | ocaml-base-compiler.4.07.0 | pinned(file:///home/travis/build/BinaryAnalysisPlatform/bap)
# path        ~/.opam/4.07.0/.opam-switch/build/bap.master
# command     ~/.opam/opam-init/hooks/sandbox.sh build ./configure --prefix=/home/travis/.opam/4.07.0 --with-cxx=`which clang++` --with-llvm-version=5.0.0git-929163d --with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config --mandir=/home/travis/.opam/4.07.0/man --enable-everything --enable-tests --disable-ida --disable-fsi-benchmark --disable-objdump
# exit-code   1
# env-file    ~/.opam/log/bap-3530-05296d.env
# output-file ~/.opam/log/bap-3530-05296d.out
<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build bap master
└─ 
╶─ No changes have been performed
```

**Comment by XVilka** [comment]
Blocked by https://github.com/alavrik/piqi/issues/62 and https://github.com/ocaml/camlp4/issues/151

**Comment by XVilka** [comment]
I checked the rest of the dependencies - they behave well for OCaml 4.08.0+, so once piqi is fixed, should be good.

**Comment by XVilka** [comment]
Note that camlp5 7.08 was released, and piqi moved to use it instead of camlp4. So please rerun the Travis, to check the build.

**Comment by ivg** [comment]
Our Travis infrastructure is still relying on opam 1.x so we don't see the updated piqi package. That only means, that we have to update our Travis CI scripts. Indeed, it doesn't make sense to test installability with 1.x

**Comment by XVilka** [comment]
I don't get the reason of 4.07 build failure:
```
<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+ /home/travis/.opam/opam-init/hooks/sandbox.sh "build" "./configure" "--prefix=/home/travis/.opam/4.07.0" "--with-cxx=`which clang++`" "--with-llvm-version=5.0.0git-929163d" "--with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config" "--mandir=/home/travis/.opam/4.07.0/man" "--enable-everything" "--enable-tests" "--disable-ida" "--disable-fsi-benchmark" "--disable-objdump" (CWD=/home/travis/.opam/4.07.0/.opam-switch/build/bap.master)
[ERROR] The compilation of bap failed at
        "/home/travis/.opam/opam-init/hooks/sandbox.sh build ./configure
        --prefix=/home/travis/.opam/4.07.0 --with-cxx=`which clang++`
        --with-llvm-version=5.0.0git-929163d
        --with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config
        --mandir=/home/travis/.opam/4.07.0/man --enable-everything
        --enable-tests --disable-ida --disable-fsi-benchmark
        --disable-objdump".
#=== ERROR while compiling bap.master =========================================#
# context     2.0.4 | linux/x86_64 | ocaml-base-compiler.4.07.0 | pinned(file:///home/travis/build/BinaryAnalysisPlatform/bap)
# path        ~/.opam/4.07.0/.opam-switch/build/bap.master
# command     ~/.opam/opam-init/hooks/sandbox.sh build ./configure --prefix=/home/travis/.opam/4.07.0 --with-cxx=`which clang++` --with-llvm-version=5.0.0git-929163d --with-llvm-config=/usr/local/clang-5.0.0/bin/llvm-config --mandir=/home/travis/.opam/4.07.0/man --enable-everything --enable-tests --disable-ida --disable-fsi-benchmark --disable-objdump
# exit-code   1
# env-file    ~/.opam/log/bap-5087-05296d.env
# output-file ~/.opam/log/bap-5087-05296d.out
<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build bap master
└─ 
╶─ No changes have been performed
The packages you requested declare the following system dependencies. Please
make sure they are installed before retrying:
    clang dejagnu libcurl4-gnutls-dev libgmp-dev libzip-dev llvm llvm-3.8-dev
    m4 time
'opam install bap -v' failed.
'unset TESTS; OPAMYES=1 export OPAMYES; set -uex; opam install bap "-v"' exited 31. Terminating with 31
The command "bash -ex .travis_install.sh" exited with 31.
cache.2
store build cache
```

**Comment by ivg** [comment]
The build fails, because the oasis wasn't updated (it also limits the version of OCaml). @gitoleg revived his old branch and managed to build everything for 4.08. While he is running his tests, here is the status update.

The 4.07 is the latest version supported by core_kernel 4.07.0, so in order to switch to 4.08 we need to switch to a newer version of core_kernel. As always, the switch is not backward compatible, so we can't support both v0.11 and v0.12 (especially, since the renamed a couple of libraries). Therefore, the solution is to switch to v0.12 and drop the support for 4.04, 4.05, and 4.06. So that now we will support only 4.07 and 4.08.

I'm totally ready to switch to 4.08+core.v0.12 even today in bap-2.0 branch. The main question is whether we should backport this to 1.x. Basically, the list of options that we have:

1) keep 1.6 on old compilers and update only 2.0
2) switch both 1.6 and 2.0 to the new compilers
3) for 1.6 into 1.6.x, 1.7.0, and 2.0.0 with the latter two using v0.12 and the first still tracking the older compiler.

Opinions, other options?

The main question is whether we should

**Comment by XVilka** [comment]
I think the third option is the sanest one, keeping one with the modern world, without the breaking old things.

**Comment by ivg** [comment]
> I think the third option is the sanest one, keeping one with the modern world, without the breaking old things.

Yep, but this implies lots of work and less clear work. If we will pursue this path, then we need to keep code compatible with different versions of OCaml (including the deprecation and removal of Pervasives and Genarray.*.map_file, which requires us to add a few shim libraries as dependencies, and use them as indirections in our code).

**Comment by XVilka** [comment]
Then the least work path

**Comment by ivg** [comment]
This branch is go to go into the master (after a merge or rebase), see [the related issue][1] for more discussion  

[1]: https://github.com/BinaryAnalysisPlatform/bap/issues/969#issuecomment-521641144

**Comment by XVilka** [comment]
Rebased against the current master

**Comment by XVilka** [comment]
I don't get why it shows while it was fixed:
```
# + /home/travis/.opam/4.07.0/bin/ocamlfind ocamlc -c -g -annot -bin-annot -short-paths -I lib/monads -I lib/bap_strings -I lib/bitvec -I lib/bitvec_sexp -I lib/bitvec_binprot -I lib/bitvec_order -I lib/knowledge -I lib/bap_core_theory -I lib/regular -I lib/graphlib -w +a-4-6-7-9-27-29-32..42-44-45-48-50-60 -package zarith -package sexplib0 -package ppx_jane -package parsexp -package ocamlgraph[...]
# File "lib/bap_lisp/bap_lisp__resolve.ml", line 115, characters 6-39:
# Error: This expression has type (var * 'weak2) list List.Or_unequal_lengths.t
#        but an expression was expected of type 'weak5 option
# Command exited with code 2.
# make: *** [build] Error 1
```
Even the line number is wrong

**Comment by XVilka** [comment]
Looks like Travis doesn't get the valid sources.

**Comment by XVilka** [comment]
Note, that the piqi [switched to sedlex](https://github.com/alavrik/piqi/commit/a7a9cd6adb6b121369fdb32b425a2c2f46194791), so no need for camlp4 and camlp5 dependency! \o/
BAP is more future-proof now. Waiting for the release now: https://github.com/alavrik/piqi/issues/61

**Comment by XVilka** [comment]
I think we also can drop these lines from the `.travis-ocaml.sh`:
```sh
  case "$OCAML_VERSION,$OPAM_VERSION" in
    3.12,1.2.2) OCAML_FULL_VERSION=3.12.1; brew install opam ;;
    4.00,1.2.2) OCAML_FULL_VERSION=4.00.1; brew install opam ;;
    4.01,1.2.2) OCAML_FULL_VERSION=4.01.0; brew install opam ;;
    4.02,1.2.2) OCAML_FULL_VERSION=4.02.3; brew install opam ;;
    4.02,1.3.0) OCAML_FULL_VERSION=4.02.3; brew install opam --HEAD ;;
```
https://github.com/BinaryAnalysisPlatform/bap/blob/master/.travis-ocaml.sh#L142
And other, similar cases too.

**Comment by XVilka** [comment]
There is still no answer from piqi author, but maybe it's possible to pin the repo for a while?

**Comment by XVilka** [comment]
Should be good to retry once https://github.com/ocaml/opam-repository/pull/14947 is merged.

**Comment by XVilka** [comment]
Still the same error. Travis seems out of sync with the git and opam.

**Comment by XVilka** [comment]
So what do we do with this one?

**Comment by gitoleg** [comment]
@XVilka 
We're currently working on the release of `bap.2.0` that should happen before the end of this month. Right after that, we'll merge this PR.

**Comment by XVilka** [comment]
So it is time, I guess?

**Comment by ivg** [comment]
yep, @gitoleg can you give us an update when we can have this merge and what would be the impact?

**Comment by gitoleg** [comment]
I don't expect any serious difficulties here since we already built this branch not a long time ago. So let's say that I will update it to the current master this week, and then I suppose we'll be ready to merge. 
Speaking about the impact we'll have: we have to drop support for three compilers: 4.04, 4.05 and 4.06, since we're going to use  `core_kernel.v0.12`

**Comment by XVilka** [comment]
I rebased it already to the current master. But Travis fails for unknown
reasons, so I guess you could check but only locally.

**Comment by ivg** [comment]
Core 0.13 has been just [released][1], it has the lower bound set to 4.08, which is probably to harsh, but we can at least investigate how much is the impact and, probably, support it in addition to the 0.12 version.


[1]: https://discuss.ocaml.org/t/ann-v0-13-release-of-jane-street-packages/4735/2
[2]: https://github.com/ocaml/opam-repository/pull/15324/files

**Comment by XVilka** [comment]
@ivg completely agree. One of the interesting features of the new library is a deriver plugin for Python accessors generation: https://github.com/janestreet/ppx_python

Since BAP has Python bindings, it might be quite useful for making them better.

File: lib/bap_future/bap_future.mli [hunk 1]
```diff
@@ -288,7 +288,29 @@ module Std : sig
     include Monad.S with type 'a t := 'a t
     include Applicative.S with type 'a t := 'a t
     module Variadic : Variadic.S with type 'a arg = 'a t
-    module Args : Applicative.Args with type 'a arg := 'a t
+
+    module Args : sig
+      type 'a arg = 'a t
```
Comment (ivg) [inline lib/bap_future/bap_future.mli]: remove the arg type from the signature, it wasn't there before

File: lib/bap_primus/bap_primus_lisp_resolve.ml [hunk 2]
```diff
@@ -107,18 +108,24 @@ let overload_macro code (s3) =
   | ((n,_,_) as c) :: cs -> List.filter_map (c::cs) ~f:(fun (m,d,bs) ->
       Option.some_if (n = m) (d,bs))
 
-let all_bindings f = 
+let all_bindings f =
   List.for_all ~f:(fun (v,x) ->
       f v.data.typ x)
 
+let zip x y =
+  let open List.Or_unequal_lengths in
```
Comment (ivg) [inline lib/bap_primus/bap_primus_lisp_resolve.ml]: There is no need for this open, as now the compiler will take the constructors from the right place.

File: lib/bap_types/bap_bitvector.ml [hunk 3]
```diff
@@ -178,15 +178,15 @@ let pp_generic
     | `this x -> fprintf ppf "%s" x
     | `base -> pp_prefix ppf
     | `auto ->
-      if Bitvec_order.(x >= (min (word 10) base))
+      if Z.Compare.(x >= (Z.min (word 10) base))
```
Comment (ivg) [inline lib/bap_types/bap_bitvector.ml]: wow, it was a bug before... not that it was breaking before, but at least the code as written was not the code as meant.

File: lib/bap_types/bap_ir.ml [hunk 4]
```diff
@@ -1558,16 +1559,17 @@ module Ir_blk = struct
       Seq.(seq jmp_t ~rev blk >>| fun x -> `Jmp x)
 
 
-  let apply_map name get map skip blk ~f =
-    if List.mem ~equal:Polymorphic_compare.equal skip name
+  let apply_map ~equal name get map skip blk ~f =
+    if List.mem ~equal skip name
```
Comment (ivg) [inline lib/bap_types/bap_ir.ml]: it is fine to use the polymorphic compare here

**Pull request merged**